### PR TITLE
Initial commit of MIRI MRS notebook

### PR DESCRIPTION
**Pull request by drlaw1558**
This PR is for the initial commit of the MIRI MRS pipeline notebook.

### CODE REVIEW

**Review by sosey — CHANGES_REQUESTED** [review]
see the review comments, there are still some changes to make so that it can run for the general pub and ci

**Comment by drlaw1558** [comment]
<p>I took at look at the JDAT format; while I don't want to follow that precisely with having a couple pages of text before the notebook gets started I have made some boldface bullets before each text block to call out more clearly what each is for.  I've changed to just point toward the notebook repository instead of the notebook file itself- it should be fairly easy to find from the top level and would be more robust against future changes.  Main aim with this is in the case that the community starts modifying and making their own versions (which we want) to have a clear place where any future users of said notebook could find any updates to the original.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325810'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
<p>I've revised this to use a 'demo_mode' argument which is TRUE by default.  In this case it overrides all of the custom user-local paths, and downloads data from Box to work on into a local directory adjacent to the notebook.  This should do what we need for CI.  Box file is currently staged in my personal Box folder- we should figure out where to move that to longer term.</p><p> </p><p>I've commented out CRDS path information too- that should be set already as part of any pipeline install, so it's only in the notebook for convenience if someone wanted to change it around.</p><p> </p><p>I've made a change to paths as well to use a default home dir instead of /Users/dlaw/ .  I'm not super keen on that since it makes the line a lot clunkier and harder to read though, especially since by definition that's the line that folks need to modify in order to run things on their own data.  Maybe there's a cleaner syntax.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325823'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
<p>Added some words.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325840'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
<p>As above; made more generic, but at the cost of readability.  Not sure that's a good trade right now given this line is only relevant when users start modifying paths anyway, but let's see.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325852'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
<p>Agreed in principle, though we also want the data set to be realistic.  In this case it's a minimal complete set of typical MRS observations.</p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325865'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
<p>Done; some worry that it could clobber a users other local plot, but likely fairly low risk.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325879'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sosey** [comment]
<p>There is already an STScI solution for using Box that does not involve personal box folders that we can use.  However, I am more worried that we should <em>not</em> use box specifically for these notebook examples. As they are directly tied to pipeline processing, and the input data may change in form and content following build deliveries in ways that would require us pulling the data from the archive again and restaging on box. I feel it would be better to pull the data from the archive directly instead. I'll suggest some alternatives. </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325823'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sosey** [comment]
<p>I made updates that pull all the data from the archive, first by query and then download. See the notebook here <a href="https://github.com/spacetelescope/jwst-pipeline-notebooks/blob/demo-example/notebooks/MIRI/JWPipeNB-MIRI-MRS.ipynb" target="_blank">https://github.com/spacetelescope/jwst-pipeline-notebooks/blob/demo-example/notebooks/MIRI/JWPipeNB-MIRI-MRS.ipynb</a></p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325823'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sosey** [comment]
<p>I added a download and query to the example notebook on the demo-example branch, we can run this through CI as part of aceptance to make sure that github actions can handle the data volume and download time</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jwst-pipeline-notebooks/pull/3/discussion/#comment-2067325865'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drlaw1558** [comment]
I just updated the MRS notebook to incorporate your approach to downloading via MAST, but tweaked a little to hide some of the demo-data stuff a bit so that it doesn't get in the way of the more regular offline-data use.  It also culls the list of files to download from >100 to just 36 (as there were many duplicates) and makes a couple other minor tweaks.

**Comment by drlaw1558** [comment]
Also need to decide whether to ingest the notebook with executed outputs (which is useful for reference to see what things should look like), or with cleared outputs (which is easier for reading the actual cells through the most basic online methods).  I'm on the fence.

**Comment by camipacifici** [comment]
In jdat_notebooks, we merge always with cleared outputs. CI runs the notebooks and generates html versions with all the outputs. Like here: https://spacetelescope.github.io/jdat_notebooks/notebooks/asdf_example/asdf_example.html

**Comment by sosey** [comment]
<p>perfect @drlaw1558. I'm sure there's more improvements to make as well before it's finalized</p>

**Comment by drlaw1558** [comment]
Adding this note here as well instead of just in an email thread: we should waive PEP8 E265 issues with these notebooks.  The purpose of the notebooks is to provide commented-out code lines that users can uncomment to change processing, so the lack of whitespace between # and commands is helpful to discriminate between real comments and potential code.

**Comment by haticekaratay** [comment]
@drlaw1558, can you rebase this branch with the main? We updated the GitHub workflows on the repo. Thanks!

**Comment by drlaw1558** [comment]
I just merged main into the branch, but looks like it's still unhappy about many things.

**Comment by haticekaratay** [comment]
@drlaw1558 Could you please rebase the branch with the main again, or merging would also work? Thanks!

**Comment by drlaw1558** [comment]
Done.

**Comment by haticekaratay** [comment]
@drlaw1558, I fixed the error in the CI runner workflow. Could I please have one more rebase? Thanks again. Hopefully, after this, the following notebooks won't need such settings.

**Comment by haticekaratay** [comment]
@rizeladiaz and @gibsongreen, the CI setup for this repo is complete. It's now ready for review.

**Comment by drlaw1558** [comment]
I've gone through the revised notebook and made a few tweaks:
- 'Configuration' was deliberately at the top to make it easier for users to find the section that they need to modify, instead of burying it 3 screens down below a ton of import cells that they *don't* need to modify.  I've moved this back.  Likewise, the list of package imports is redundant and takes substantial unnecessary screen real estate.
- We should assume that users have installed the JWST pipeline correctly and provide optional CRDS lines in case they have not, rather than setting CRDS paths by default and requiring users to comment them out (I can see this causing issues for many folks)
- I don't think we should repeat the pipeline version in comments a few lines below the Pipeline Version entry; this will very easily get mismatched in future.  Changed the phrasing.
- Removed some redundant comments (e.g., two lines saying what a cell configures)

**Comment by camipacifici** [comment]
Doesn't CI need the CRDS configuration?

**Comment by drlaw1558** [comment]
> Doesn't CI need the CRDS configuration?

I would hope the CI system is set up such that the CRDS framework is already installed, rather than adding yet more code that users need to remove before they can use the notebook.

**Comment by haticekaratay** [comment]
> > Doesn't CI need the CRDS configuration?
> 
> I would hope the CI system is set up such that the CRDS framework is already installed, rather than adding yet more code that users need to remove before they can use the notebook.

Hi David,
I haven't been keeping up with the conversation about this notebook. If you are talking about setting up the environment variables for the CRDS and installing relevant data, we should incorporate that into the notebook workflow. Since most of our users are astronomers, we aim to ensure that they can run this notebook locally without needing any additional information beyond what is provided here. They should be able to run the notebook without assuming knowledge of any unknown dependencies. Therefore, I suggest that we include the steps for setting up these environment variables, if that's what you are referencing.

**Comment by drlaw1558** [comment]
Hi Hatice-

In many cases I'd agree, but as the purpose of this repository is to contain notebooks that run the pipeline, we should work on the assumption that they've successfully followed the instructions at https://github.com/spacetelescope/jwst (and elsewhere) on how to install the pipeline.  One possibility would be to add an if/else catch- if the required CRDS information is not set then set it, but I wouldn't want to be overwriting this for users who installed properly.

What I'm unclear on though is whether this is actually what's stopping CI from being successful; it looked like there were still CI crashes even in the version that explicitly set these CRDS paths?

**Comment by rizeladiaz** [comment]
The issue beyond setting CRDS seems to be a mismatch between numpy 2.0 and matplotlib. The requirements.txt file should be:

numpy>=2
crds
jwst>=1.13.3
astroquery
matplotlib>=3.9.0

I can try that and run the test again to see if that works before pushing the changes again.

Regarding CRDS. The README of the noteboks should clearly state that users need to set up these variables ahead of time. This, however, does not address the issue we have now.  I would suggest to declare them in the notebook but add a big warning to users about commenting that line if their environmet variables are already set up. I don't think that adding an  "if else" statement would be good idea.

**Comment by sosey** [comment]
JWST pipeline is requiring numpy <2.0 so the requirements should respect that at this point....
https://github.com/spacetelescope/jwst/blob/master/pyproject.toml#L27

**Comment by drlaw1558** [comment]
I think I've got the dependencies all adjusted and I'm no longer getting immediate execution crashes, but it sure seems to be taking a long time in the 'Validate notebooks' step.  What is this step actually doing, as it's prior to notebook execution?

**Comment by drlaw1558** [comment]
Looks like the latest error is that CI is running out of disk space with just 71 MB.  Not sure what it started at?  Where is the actual CI server based, is it shared between different projects, etc?  We either need to clean up the machine and/or substantially increase the available space- even a minimum typical JWST data set is going to run to a couple of GB.

**Review by sosey — APPROVED** [review]
I'm assuming that the changes have been made with the follow-on reviews

File: notebooks/MIRI/JWPipeNB-MIRI-MRS.ipynb
Comment (rizeladiaz) [inline notebooks/MIRI/JWPipeNB-MIRI-MRS.ipynb:1]: I had reviewed this file before. The information never made it to this repository, so I am adding it now 

Review changes:
- Added to requirements.txt file ‘astroquery'

First section:
 - Changed text 4-pt dither to 4-point dither to be consistent with JWST User documentation
 - Defined acronyms MIRI, MRS, JWST, MAST
 - removed unnecessary caps e.g. Observation
 - Made  link to section "Configuration" instead of just having it with quotes

Other sections:
- Added to heading a section documenting what JWST pipeline version and CRDS context were used for the notebook.
- Provide links to important information when needed; e.g. Astroquery user guide
- Changed all directory setup and path + filename definitions to use the "os"  module to adjust to OS.
- Added pathlib.Path to represent filesystem paths with semantics appropriate for different operating systems (needed in astro query call)
- Changed to use “with" to open files for proper acquisition and release of resources
- Removed trailing whitespaces 
- Added two spaces before inline comments
- Reduced some line lengths
- spec3 changed to only include background files for selected channel and band

As stated by David, commented command lines were left as is, no space between "#' and start of the command to distinguish them from real comments.
- changed section headings from  highlighted format to regular markdown and only leave important considerations in highlighted format.
- Defined convenience function section to add one function: select files for specific channel/band. This avoid repeating it through the whole notebook

**Review by rizeladiaz — APPROVED** [review]
Agree with changes

**Review by sosey — CHANGES_REQUESTED** [review]
Blocking this from merge until further discussions happen.

**Review by sosey — APPROVED** [review]
approving for merge after discussion with the about the benefits of having an example notebooks in the main branch for ins to start building off of. Testing to understand larger runners will continue in parallel

**Comment by rizeladiaz** [comment]
Merging pull request to make notebook available to Instrument Branches

**Pull request merged**
